### PR TITLE
fix voxelizer

### LINKED ISSUE #598: error in voxelization
Fix this: 
image

### PR DESCRIPTION
**Pull request by stuikomma**
Voxel that had both a negative and a positive normal at the same z level would get filled wrongly if the former was voxelized first. Now the positive normals take precedence over the other ones.

### CODE REVIEW

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 1]
```diff
@@ -178,13 +187,23 @@ module.exports = class Voxelizer
...
 				roundA = @_roundVoxelSpaceToVoxel a
+				roundB = @_roundVoxelSpaceToVoxel b
+
 				if roundA.x is x and roundA.y is y and roundA.z is z
-					# aIsInVoxel
-					roundB = @_roundVoxelSpaceToVoxel b
-					if roundB.x is x and roundB.y is y and roundB.z is z
-						# bIsInVoxel
-						return Math.max a.z, b.z
+					aIsInVoxel = true
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: `aIsInVoxel = roundA.x is x and roundA.y is y and roundA.z is z`

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 2]
```diff
@@ -178,13 +187,23 @@ module.exports = class Voxelizer
...
+				else
+					aIsInVoxel = false
+				if roundB.x is x and roundB.y is y and roundB.z is z
+					bIsInVoxel = true
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: `bIsInVoxel = roundB.x is x and roundB.y is y and roundB.z is z`

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 3]
```diff
@@ -178,13 +187,23 @@ module.exports = class Voxelizer
...
+				else
+					bIsInVoxel = false
 
+				if aIsInVoxel && bIsInVoxel
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: https://github.com/style-guides/CoffeeScript#relational-operators

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 4]
```diff
@@ -227,12 +246,13 @@ module.exports = class Voxelizer
...
 				}
 
-			_setVoxel: ({x: x, y: y, z: z}, zValue, direction, grid) ->
+			_setVoxel: ({x: x, y: y, z: z}, zValue, direction, grid, floatDelta) ->
 				grid[x] = [] unless grid[x]
 				grid[x][y] = [] unless grid[x][y]
 				oldValue = grid[x][y][z]
 				if oldValue
-					if oldValue.dir is 0 or (zValue > oldValue.z and direction isnt 0)
+					if oldValue.dir is 0 or (direction isnt 0 and zValue > oldValue.z) or
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: Maybe it is time to extract some of those conditions and add comments.

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 5]
```diff
@@ -16,13 +18,20 @@ module.exports = class Voxelizer
...
 			if message.state is 'progress'
 				progressCallback message.progress
 			else # if state is 'finished'
 				@resolve grid: @voxelGrid, gridPOJO: message.data
 
+		@worker = @_getWorker()
+		@worker.voxelize(
+			voxelSpaceModel
+			lineStepSize
+			floatDelta
```
Comment (Steditor) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: Should probably be `@floatDelta`

Comment (stuikomma) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: No, because it's not within the class. Should it?

Comment (Steditor) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: It already is: The `voxelize` method (not `worker.voxelize`) is within the class. (Which is also why you are able to call `@worker` btw)
I cannot find any definition of `floatDelta` here, only `@floatDelta`.

Comment (stuikomma) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: I meant that the definition of floatDelta isn't. It's in line 3.

Comment (Steditor) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: Ah, I see. I was thrown off by the usage of `@floatDelta` below - but it's the worker there...

Comment (stuikomma) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: Would you prefer to have it in within the class?  I think there it would be less visible.

Comment (Steditor) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: No - it's perfectly fine the way it is!

**Comment by stuikomma** [comment]
image

**Comment by stuikomma** [comment]
This issue cannot be fixed with our algorithm since it is a very specific error related to this model.

The bottom and top voxel are voxelized as part of the faces with vertices at `x = 1.499999999998`. This leads to the creation of a voxel at `x = 1`. The faces in the middle where the model should be closed again only have vertices at `x = 1.5` which gets rounded upt to `x = 2`.

Actually there should be a normal that points upwards in between but that is below our threshold value (`zTolerance = 0.01`). If this threshold is lowered other errors occur.

With the old algorithm it worked because these voxel were regarded as 'level' (i.e. neither up or down), which would lead to holes in models elsewhere.

**Comment by Steditor** [comment]
Now also fixes #598 

From my perspective: :boat: - any issues with other models you would like to test?

**Comment by Steditor** [comment]
There are missing voxels inside the ballista, therefore: :boat: :arrow_right: :put_litter_in_its_place:

**Comment by AtiX** [comment]
- [ ] fix missing voxels inside ballista

**Comment by stuikomma** [comment]
:put_litter_in_its_place::arrow_right::boat::question:

File: src/plugins/newBrickator/pipeline/VolumeFiller.coffee [hunk 6]
```diff
@@ -79,8 +76,8 @@ module.exports = class VolumeFiller
 				return
 
 			_setVoxels: (grid, x, y, zs, voxelData) ->
-				for z in zs
-					@_setVoxel grid, x, y, z, voxelData
+				while zs.length > 0
+					@_setVoxel grid, x, y, zs.pop(), voxelData
```
Comment (adius) [inline src/plugins/newBrickator/pipeline/VolumeFiller.coffee]: Maybe:
```
while zValues = zs.pop()
    @_setVoxel grid, x, y, zValues, voxelData
```

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 7]
```diff
@@ -154,21 +166,27 @@ module.exports = class Voxelizer
...
-					(oldVoxel.y != currentVoxel.y) or
-					(oldVoxel.z != currentVoxel.z)
-						z = @_getGreatestZInVoxel a, b, currentVoxel
-						@_setVoxel currentVoxel, z, direction, grid
+					unless @_isOnVoxelBorder currentGridPosition
+						oldVoxel = currentVoxel
+						currentVoxel = @_roundVoxelSpaceToVoxel currentGridPosition
+						if (oldVoxel.x != currentVoxel.x) or
+						(oldVoxel.y != currentVoxel.y) or
+						(oldVoxel.z != currentVoxel.z)
```
Comment (adius) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: isnt instad of !=

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 8]
```diff
@@ -178,61 +196,64 @@ module.exports = class Voxelizer
...
-						# bIsInVoxel
-						return Math.max a.z, b.z
+				roundB = @_roundVoxelSpaceToVoxel b
+
+				aIsInVoxel = roundA.x is x and roundA.y is y and roundA.z is z
+				bIsInVoxel = roundB.x is x and roundB.y is y and roundB.z is z
 
+				if aIsInVoxel and bIsInVoxel
+					return Math.max a.z, b.z
+				if aIsInVoxel && a.z > b.z
```
Comment (adius) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: and instead of &&

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 9]
```diff
@@ -178,61 +196,64 @@ module.exports = class Voxelizer
...
+					return a.z
+				if bIsInVoxel && b.z > a.z
```
Comment (adius) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: and instead of &&

File: src/plugins/newBrickator/pipeline/HullVoxelizer.coffee [hunk 10]
```diff
@@ -16,13 +19,21 @@ module.exports = class Voxelizer
 
 		voxelSpaceModel = @_getOptimizedVoxelSpaceModel optimizedModel, options
 
-		@worker = @_getWorker()
-		@worker.voxelize voxelSpaceModel, lineStepSize, (message) =>
+		callback = (message) =>
```
Comment (adius) [inline src/plugins/newBrickator/pipeline/HullVoxelizer.coffee]: what callback? Please think of a better name

**Pull request merged**